common/src/main/java/com/aliyun/emr/rss/common/network/util/MemoryTracker.java

```diff
@@ -0,0 +1,196 @@
+    assert lowRatio > 0 && lowRatio < 1;
+    assert directMemoryCriticalRatio > 0 && directMemoryCriticalRatio < 1;
+    high = (long) (maxDirectorMemory * highRatio);
+    low = (long) (maxDirectorMemory * lowRatio);
+    directMemoryCritical = (long) (maxDirectorMemory * directMemoryCriticalRatio);
+    assert high > 0;
+    assert low > 0;
+    assert directMemoryCritical > 0;
+
+    scheduledExecutorService.scheduleWithFixedDelay(() -> {
+      try {
+        if (needFreeMemory()) {
+          logger.info("Trigger storage memory on-pressure action");
+          listeners.forEach(MemoryTrackerListener::onMemoryPressure);
+        }
+      } catch (Exception e) {
+        logger.error("Storage memory release on high pressure with error");
+      }
+      if (logger.isDebugEnabled()) {
+        if (storageMemoryTracker.get() != 0) {
+          logger.debug("Storage used memory :{} MB",
+            toMb(storageMemoryTracker.get()));
+          if (directMemoryIndicator != null) {
+            logger.debug("Track all direct memory usage :{}/{} ({}/{} MB)",
+              directMemoryIndicator.get(), maxDirectorMemory, toMb(directMemoryIndicator.get()),
+              toMb(maxDirectorMemory));
+          }
+        }
+      }
+    }, interval, interval, TimeUnit.MILLISECONDS);
+
+    reportService.scheduleWithFixedDelay(() -> {
+      if (storageMemoryTracker.get() != 0) {
+        long writeSpeed = writeSpeedCounter.sumThenReset();
+        long flushSpeed = flushSpeedCounter.sumThenReset();
+        logger.info("Storage used memory :{} MB",
+          toMb(storageMemoryTracker.get()));
+        logger.info("Storage memory write speed :{} MB/s",
+          toMb(writeSpeed) / 10);
+        logger.info("Storage memory flush speed :{} MB/s",
+          toMb(flushSpeed) / 10);
+        if (directMemoryIndicator != null) {
+          logger.info("Track all direct memory usage :{}/{}",
+            toMb(directMemoryIndicator.get()), toMb(maxDirectorMemory));
+        }
+      }
+    }, 10, 10, TimeUnit.SECONDS);
+
+    logger.info("Storage memory tracker initialized with :  " +
+                  "\n max direct memory : {} ({} MB)" +
+                  "\n direct memory critical : {} ({} MB)" +
+                  "\n high water mark : {} ({} MB)" +
+                  "\n low water mark : {} ({} MB)", maxDirectorMemory, toMb(maxDirectorMemory),
+      directMemoryCritical, toMb(directMemoryCritical), high, toMb(high), low, toMb(low));
+
+    initDirectMemoryIndicator();
+  }
+
+  private double toMb(long bytes) {
+    return bytes / 1024.0 / 1024.0;
+  }
+
+  private void initDirectMemoryIndicator() {
+    Field field = null;
+    Field[] result = PlatformDependent.class.getDeclaredFields();
+    for (Field tf : result) {
+      if ("DIRECT_MEMORY_COUNTER".equals(tf.getName())) {
+        field = tf;
+      }
+    }
+    field.setAccessible(true);
+    try {
+      directMemoryIndicator = ((AtomicLong) field.get(PlatformDependent.class));
```

Comment:
Not sure if directMemoryIndicator always returns the current value of DIRECT_MEMORY_COUNTER, is this tested?